File: examples/unreal-client/AgentEClient.cpp
/**
 * AgentE Unreal Engine Client — Implementation
 *
 * See AgentEClient.h for setup instructions.
 * TODO: Replace BuildStateJson() with your actual economy data.
 */

#include "AgentEClient.h"
#include "HttpModule.h"
#include "Interfaces/IHttpResponse.h"

UAgentEClient::UAgentEClient()
{
    PrimaryComponentTick.bCanEverTick = false;
}

void UAgentEClient::BeginPlay()
{
    Super::BeginPlay();
    UE_LOG(LogTemp, Log, TEXT("[AgentE] Client initialized, server: %s"), *ServerUrl);
}

// ─── Game Loop Integration ──────────────────────────────────────────────────

void UAgentEClient::OnGameTick()
{
    TickCounter++;
    if (TickCounter % TickInterval == 0)
    {
        SendTick();
    }
}

// ─── HTTP Communication ─────────────────────────────────────────────────────

void UAgentEClient::SendTick()
{
    FString StateJson = BuildStateJson();
    FString Body = FString::Printf(TEXT("{\"state\":%s}"), *StateJson);

    TSharedRef<IHttpRequest, ESPMode::ThreadSafe> Request =
        FHttpModule::Get().CreateRequest();

    Request->SetURL(ServerUrl + TEXT("/tick"));
    Request->SetVerb(TEXT("POST"));
    Request->SetHeader(TEXT("Content-Type"), TEXT("application/json"));
    Request->SetContentAsString(Body);
    Request->OnProcessRequestComplete().BindUObject(
        this, &UAgentEClient::HandleTickResponse);

    Request->ProcessRequest();
}

void UAgentEClient::CheckHealth()
{
    TSharedRef<IHttpRequest, ESPMode::ThreadSafe> Request =
        FHttpModule::Get().CreateRequest();

    Request->SetURL(ServerUrl + TEXT("/health"));
    Request->SetVerb(TEXT("GET"));
    Request->OnProcessRequestComplete().BindLambda(
        [](FHttpRequestPtr, FHttpResponsePtr Response, bool bSuccess) {
            if (bSuccess && Response.IsValid())
            {
                UE_LOG(LogTemp, Log, TEXT("[AgentE] Health: %s"),
                    *Response->GetContentAsString());
            }
        });

    Request->ProcessRequest();
}

// ─── State Building ─────────────────────────────────────────────────────────
// TODO: Replace this with your actual economy data

FString UAgentEClient::BuildStateJson()
{
    // Build JSON from your game's economy state
    // In production, use FJsonObject for proper serialization

    return FString::Printf(TEXT(R"({
        "tick": %d,
        "roles": ["role_a", "role_b", "role_c"],
        "resources": ["resource_x", "resource_y"],
        "currencies": ["currency_a"],
        "agentBalances": {
            "agent_1": { "currency_a": 150 },
            "agent_2": { "currency_a": 80 }
        },
        "agentRoles": {
            "agent_1": "role_a",
            "agent_2": "role_b"
        },
        "agentInventories": {
            "agent_1": { "resource_x": 2 },
            "agent_2": { "resource_y": 5 }
        },
        "marketPrices": {
            "currency_a": { "resource_x": 15, "resource_y": 50 }
        },
        "recentTransactions": []
    })"), TickCounter);
}

// ─── Response Handling ──────────────────────────────────────────────────────

void UAgentEClient::HandleTickResponse(FHttpResponsePtr Response, bool bSuccess)
{
    if (!bSuccess || !Response.IsValid())
    {
        UE_LOG(LogTemp, Warning, TEXT("[AgentE] Tick request failed"));
        return;
    }

    TSharedPtr<FJsonObject> JsonObject;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(
        Response->GetContentAsString());

    if (!FJsonSerializer::Deserialize(Reader, JsonObject) || !JsonObject.IsValid())
    {
        UE_LOG(LogTemp, Warning, TEXT("[AgentE] Failed to parse response"));
        return;
    }

    // Update health
    LastHealth = JsonObject->GetIntegerField(TEXT("health"));
    UE_LOG(LogTemp, Log, TEXT("[AgentE] Health: %d/100"), LastHealth);

    // Process adjustments
    const TArray<TSharedPtr<FJsonValue>>* Adjustments;
    if (JsonObject->TryGetArrayField(TEXT("adjustments"), Adjustments))
    {
        for (const auto& AdjValue : *Adjustments)
        {
            TSharedPtr<FJsonObject> Adj = AdjValue->AsObject();
            if (Adj.IsValid())
            {
                FString Key = Adj->GetStringField(TEXT("key"));
                float Value = Adj->GetNumberField(TEXT("value"));

                UE_LOG(LogTemp, Log, TEXT("[AgentE] Adjust %s -> %f"), *Key, Value);
                OnAdjustmentReceived.Broadcast(Key, Value);
            }
        }
    }

    // Process alerts
    const TArray<TSharedPtr<FJsonValue>>* Alerts;
    if (JsonObject->TryGetArrayField(TEXT("alerts"), Alerts))
    {
        for (const auto& AlertValue : *Alerts)
        {
            TSharedPtr<FJsonObject> Alert = AlertValue->AsObject();
            if (Alert.IsValid())
            {
                FString Principle = Alert->GetStringField(TEXT("principle"));
                FString Name = Alert->GetStringField(TEXT("name"));
                int32 Severity = Alert->GetIntegerField(TEXT("severity"));

                OnAlertReceived.Broadcast(Principle, Name, Severity);
            }
        }
    }
}

File: examples/unity-client/AgentEClient.cs
/**
 * AgentE Unity Client
 *
 * Drop this script onto a GameObject in your scene.
 * It sends economy snapshots to the AgentE server every N ticks
 * and applies the returned parameter adjustments.
 *
 * Setup:
 *   1. Start AgentE server: npx @agent-e/server --port 3000
 *   2. Attach this script to a GameObject
 *   3. Implement BuildState() with your actual economy data
 *   4. Implement ApplyAdjustment() to modify your economy params
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

public class AgentEClient : MonoBehaviour
{
    [Header("Server")]
    [SerializeField] private string serverUrl = "http://localhost:3000";
    [SerializeField] private int tickInterval = 5; // Send every N game ticks

    private int tickCounter = 0;

    // ─── Game Loop Integration ──────────────────────────────────────────

    /// Call this from your game tick/update loop
    public void OnGameTick()
    {
        tickCounter++;
        if (tickCounter % tickInterval == 0)
        {
            StartCoroutine(SendTick());
        }
    }

    // ─── HTTP Communication ─────────────────────────────────────────────

    private IEnumerator SendTick()
    {
        string stateJson = BuildState();
        string body = "{\"state\":" + stateJson + "}";

        using (UnityWebRequest req = new UnityWebRequest(serverUrl + "/tick", "POST"))
        {
            byte[] bodyRaw = Encoding.UTF8.GetBytes(body);
            req.uploadHandler = new UploadHandlerRaw(bodyRaw);
            req.downloadHandler = new DownloadHandlerBuffer();
            req.SetRequestHeader("Content-Type", "application/json");

            yield return req.SendWebRequest();

            if (req.result == UnityWebRequest.Result.Success)
            {
                HandleTickResponse(req.downloadHandler.text);
            }
            else
            {
                Debug.LogWarning($"[AgentE] Tick failed: {req.error}");
            }
        }
    }

    public IEnumerator CheckHealth()
    {
        using (UnityWebRequest req = UnityWebRequest.Get(serverUrl + "/health"))
        {
            yield return req.SendWebRequest();
            if (req.result == UnityWebRequest.Result.Success)
            {
                Debug.Log($"[AgentE] Health: {req.downloadHandler.text}");
            }
        }
    }

    // ─── State Building ─────────────────────────────────────────────────
    // TODO: Replace this with your actual economy data

    private string BuildState()
    {
        // Build the JSON state snapshot from your game
        // This is a simplified example — use JsonUtility or Newtonsoft.Json
        // for production code

        return @"{
            ""tick"": " + tickCounter + @",
            ""roles"": [""role_a"", ""role_b"", ""role_c""],
            ""resources"": [""resource_x"", ""resource_y""],
            ""currencies"": [""currency_a""],
            ""agentBalances"": {
                ""agent_1"": { ""currency_a"": 150 },
                ""agent_2"": { ""currency_a"": 80 }
            },
            ""agentRoles"": {
                ""agent_1"": ""role_a"",
                ""agent_2"": ""role_b""
            },
            ""agentInventories"": {
                ""agent_1"": { ""resource_x"": 2 },
                ""agent_2"": { ""resource_y"": 5 }
            },
            ""marketPrices"": {
                ""currency_a"": { ""resource_x"": 15, ""resource_y"": 50 }
            },
            ""recentTransactions"": []
        }";
    }

    // ─── Response Handling ───────────────────────────────────────────────

    private void HandleTickResponse(string json)
    {
        // Parse the response — use JsonUtility or Newtonsoft.Json in production
        var response = JsonUtility.FromJson<TickResponse>(json);

        Debug.Log($"[AgentE] Health: {response.health}/100");

        // TODO: Apply each adjustment to your economy parameters
        // foreach (var adj in response.adjustments)
        // {
        //     ApplyAdjustment(adj.key, adj.value);
        // }
    }

    /// TODO: Implement this to modify your game's economy parameters
    private void ApplyAdjustment(string paramKey, float value)
    {
        // Example:
        // if (paramKey == "your_cost_param") economyParams.costParam = value;
        // if (paramKey == "your_yield_param") economyParams.yieldParam = value;
        Debug.Log($"[AgentE] Adjust {paramKey} → {value}");
    }

    // ─── Response Types ─────────────────────────────────────────────────

    [Serializable]
    public class TickResponse
    {
        public Adjustment[] adjustments;
        public Alert[] alerts;
        public int health;
    }

    [Serializable]
    public class Adjustment
    {
        public string key;
        public float value;
    }

    [Serializable]
    public class Alert
    {
        public string principle;
        public string name;
        public int severity;
    }
}
